File: week4_C/q9.c
#include <stdio.h>
#include <stdlib.h>
struct Node { int data; struct Node* next; };
int getCount(struct Node* head) {
    int count = 0;
    while (head) { count++; head = head->next; }
    return count;
}
int getIntersection(int d, struct Node* head1, struct Node* head2) {
    struct Node* current1 = head1;
    struct Node* current2 = head2;
    for (int i = 0; i < d; i++) current1 = current1->next;
    while (current1 && current2) {
        if (current1 == current2) return current1->data;
        current1 = current1->next;
        current2 = current2->next;
    }
    return -1;
}
int main() {
    struct Node* newNode;
    struct Node* head1 = malloc(sizeof(struct Node));
    struct Node* head2 = malloc(sizeof(struct Node));
    head1->data = 10; head1->next = NULL;
    head2->data = 3; head2->next = NULL;
    newNode = malloc(sizeof(struct Node));
    newNode->data = 15; newNode->next = NULL;
    head1->next = newNode;
    head2->next = newNode;
    printf("%d", getIntersection(0, head1, head2));
}

File: week4/q9.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
def get_count(head):
    count = 0
    while head:
        count += 1
        head = head.next
    return count
def get_intersection(d, head1, head2):
    for _ in range(d):
        head1 = head1.next
    while head1 and head2:
        if head1 == head2:
            return head1.data
        head1 = head1.next
        head2 = head2.next
    return -1
head1 = Node(10)
head2 = Node(3)
new_node = Node(15)
head1.next = new_node
head2.next = new_node
print(get_intersection(0, head1, head2))